onnxruntime/core/providers/coreml/model/model.h

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+
+#include "core/common/status.h"
+#include "core/platform/ort_mutex.h"
+
+namespace onnxruntime {
+namespace coreml {
+
+class Execution;
+
+struct OnnxTensorInfo {
+  const int32_t data_type;  // Uses TensorProto::DataType
+  const std::vector<int64_t> shape;
+};
+
+struct OnnxTensorData {
+  OnnxTensorInfo tensor_info;
+  void* buffer{nullptr};
+};
+
+class Model {
+ public:
+  Model(const std::string& path);
+  ~Model();
+  Model(const Model&) = delete;
+  Model& operator=(const Model&) = delete;
+
+  onnxruntime::common::Status LoadModel();
+  onnxruntime::common::Status Predict(const std::unordered_map<std::string, OnnxTensorData>& inputs,
+                                      const std::unordered_map<std::string, OnnxTensorData>& outputs);
+
+  bool IsScalarOutput(const std::string& output_name) const;
+  void SetScalarOutputs(std::unordered_set<std::string>&& scalar_outputs) {
+    scalar_outputs_ = std::move(scalar_outputs);
+  }
+
+  // Mutex for exclusive lock to this model object
+  OrtMutex& GetMutex() { return mutex_; }
+
+  // Input and output names in the onnx model's order
+  const std::vector<std::string>& GetInputs() const { return inputs_; }
+  void SetInputs(std::vector<std::string>&& inputs) { inputs_ = std::move(inputs); }
+
+  const std::vector<std::string>& GetOutputs() const { return outputs_; }
+  void SetOutputs(std::vector<std::string>&& outputs) { outputs_ = std::move(outputs); }
+
+  void SetInputOutputInfo(std::unordered_map<std::string, OnnxTensorInfo>&& input_output_info) {
+    input_output_info_ = std::move(input_output_info);
+  }
```

Comment:
Can the inputs, outputs and input_output_info be provided when the model is constructed? In general you want to fully initialize a class when constructed instead of bit by bit, especially given we have a model builder class which could assemble the pieces and create the Model instance when it has those.

Reply:
Moved these constructing fuctions into private and made the ModelBuilder a friend class, the reason I don't want to put all these into the ctor is that we may add more data members later, (NNAPI's Model class has tons of members which were added piece by piece)

---
In reply to: [563464145](https://github.com/microsoft/onnxruntime/pull/6392#discussion_r563464145) [](ancestors = 563464145)